sql/dd/impl/bootstrap/bootstrapper.cc

```diff
@@ -82,11 +82,30 @@ namespace {
 // Initialize recovery in the DDSE.
 bool DDSE_dict_recover(THD *thd, dict_recovery_mode_t dict_recovery_mode,
                        uint version) {
-  handlerton *ddse = get_dd_engine(thd);
-  if (ddse->dict_recover == nullptr) return true;
+  handlerton *ddse = nullptr;
 
+  if (opt_initialize) {
+    ddse = get_dd_engine(thd);
+  } else {
+    ddse = ha_resolve_by_legacy_type(
+        thd, bootstrap::DD_bootstrap_ctx::instance().get_actual_dd_engine());
+  }
+
+  if (ddse->dict_recover == nullptr) return true;
   bool error = ddse->dict_recover(dict_recovery_mode, version);
+  if (error) return error;
 
+  /*
+    Always call innodb handler API unless disabled
+    innobase_dict_recover will initialze innodb services, which is required
+    for DDSE dd change
```

Comment:
I think code from https://github.com/facebook/mysql-5.6/pull/1357/files#diff-018eea8366c7c269ff17becee828ebbd4d3e029b47a766d199689119732e8f8bR89 has to be added there for the error case